pkg/pquerier/select.go

```diff
@@ -0,0 +1,435 @@
+		}
+	}
+
+	for i := 0; i < s.workers; i++ {
+		close(s.requestChannels[i])
+	}
+
+	// wait for Go routines to complete
+	s.wg.Wait()
+	close(s.errorChannel)
+
+	// return first error
+	err = <-s.errorChannel
+	if err != nil {
+		return nil, err
+	}
+
+	if len(s.frameList) > 0 {
+		s.totalColumns = s.frameList[0].Len()
+	}
+
+	frameIter := NewFrameIterator(s)
+	return frameIter, nil
+}
+
+func (s *selectQueryContext) metricsAggregatesToString(metric string) (string, bool) {
+	var result strings.Builder
+	specs := s.columnsSpecByMetric[metric]
+	specsNum := len(specs)
+	if specsNum == 0 {
+		return "", false
+	}
+
+	var requestedRawColumn bool
+	result.WriteString(specs[0].function.String())
+	for i := 1; i < specsNum; i++ {
+		if specs[i].function.String() == "" {
+			requestedRawColumn = true
+		} else {
+			result.WriteString(",")
+			result.WriteString(specs[i].function.String())
+		}
+	}
+
+	return result.String(), requestedRawColumn && result.Len() > 0
+}
+
+// Query a single partition
+func (s *selectQueryContext) queryPartition(partition *partmgr.DBPartition) ([]*partQuery, error) {
+	var queries []*partQuery
+	var err error
+
+	mint, maxt := partition.GetPartitionRange()
+	step := s.step
+
+	if s.maxt < maxt {
+		maxt = s.maxt
+	}
+
+	if s.mint > mint {
+		mint = s.mint
+	}
+
+	for metric := range s.columnsSpecByMetric {
+		var aggregationParams *aggregate.AggregationParams
+		functions, requestAggregatesAndRaw := s.metricsAggregatesToString(metric)
+
+		//if functions == "" && step > 0 && step >= partition.RollupTime() && partition.AggrType().HasAverage() {
+		//	functions = "avg"
+		//}
+
+		// Check whether there are aggregations to add and aggregates aren't disabled
+		if functions != "" && !s.disableAllAggr {
+
+			if step > partition.RollupTime() && s.disableClientAggr {
+				step = partition.RollupTime()
+			}
+
+			params, err := aggregate.NewAggregationParams(functions,
+				"v",
+				partition.AggrBuckets(),
+				step,
+				partition.RollupTime(),
+				s.windows)
+
+			if err != nil {
+				return nil, err
+			}
+			aggregationParams = params
+
+		}
+
+		newQuery := &partQuery{mint: mint, maxt: maxt, partition: partition, step: step}
+		if aggregationParams != nil {
+			newQuery.preAggregated = aggregationParams.CanAggregate(partition.AggrType())
+			if newQuery.preAggregated || !s.disableClientAggr {
+				newQuery.aggregationParams = aggregationParams
+			}
+		}
+
+		err = newQuery.getItems(s, metric, requestAggregatesAndRaw)
+		queries = append(queries, newQuery)
+	}
+
+	return queries, err
+}
+
+func (s *selectQueryContext) startCollectors() error {
+
+	s.requestChannels = make([]chan *qryResults, s.workers)
+	s.errorChannel = make(chan error, s.workers)
+
+	for i := 0; i < s.workers; i++ {
+		newChan := make(chan *qryResults, 1000)
+		s.requestChannels[i] = newChan
+
+		// Increment the WaitGroup counter.
+		s.wg.Add(1)
+		go func(index int) {
+			mainCollector(s, s.requestChannels[index])
+		}(i)
+	}
+
+	return nil
```

Comment:
should return errors chanel

Reply:
the error channel is part of the query context so it's accessible there anyway